File: _DEMO/DemoManager.cs
using UnityEngine;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace DEMO
{
    /// <summary>
    /// Central manager for demo file operations (save, load, browse, delete).
    /// Provides a unified interface for all demo management tasks.
    /// </summary>
    public class DemoManager : MonoBehaviour
    {
        [Header("Settings")]
        [SerializeField] private string demosDirectory = "Demos";
        [SerializeField] private int maxDemosToKeep = 100;
        [SerializeField] private long maxStorageBytes = 500 * 1024 * 1024;  // 500 MB

        [Header("Auto-Delete")]
        [SerializeField] private bool autoDeleteOldDemos = true;
        [SerializeField] private int demoRetentionDays = 30;

        private string _fullDemosPath;
        private List<DemoInfo> _demoCache;
        private DemoRecorder _recorder;
        private DemoPlayer _player;

        public string DemosDirectory => _fullDemosPath;
        public int DemoCount => _demoCache?.Count ?? 0;
        public long TotalStorageUsed => CalculateTotalStorage();

        public event Action<DemoInfo> OnDemoSaved;
        public event Action<DemoInfo> OnDemoDeleted;
        public event Action OnDemoCacheUpdated;

        private void Awake()
        {
            _fullDemosPath = Path.Combine(Application.persistentDataPath, demosDirectory);
            _demoCache = new List<DemoInfo>();
        }

        private void Start()
        {
            InitializeDemosDirectory();
            RefreshDemoCache();

            _recorder = FindAnyObjectByType<DemoRecorder>();
            _player = FindAnyObjectByType<DemoPlayer>();
        }

        /// <summary>
        /// Initialize the demos directory if it doesn't exist.
        /// </summary>
        private void InitializeDemosDirectory()
        {
            if (!Directory.Exists(_fullDemosPath))
            {
                Directory.CreateDirectory(_fullDemosPath);
                Debug.Log($"[DemoManager] Created demos directory: {_fullDemosPath}");
            }
        }

        /// <summary>
        /// Refresh the demo cache by scanning the directory.
        /// </summary>
        public void RefreshDemoCache()
        {
            _demoCache.Clear();

            if (!Directory.Exists(_fullDemosPath))
                return;

            string[] files = Directory.GetFiles(_fullDemosPath, "*.demo");

            foreach (string file in files)
            {
                try
                {
                    DemoInfo info = LoadDemoInfo(file);
                    if (info != null)
                    {
                        _demoCache.Add(info);
                    }
                }
                catch (Exception e)
                {
                    Debug.LogWarning($"[DemoManager] Failed to load demo info for {file}: {e.Message}");
                }
            }

            // Sort by date (newest first)
            _demoCache = _demoCache.OrderByDescending(d => d.timestamp).ToList();

            OnDemoCacheUpdated?.Invoke();

            Debug.Log($"[DemoManager] Refreshed cache: {_demoCache.Count} demos");
        }

        /// <summary>
        /// Load demo metadata from file.
        /// </summary>
        private DemoInfo LoadDemoInfo(string filePath)
        {
            FileInfo fileInfo = new FileInfo(filePath);

            // Read just the header to get metadata
            try
            {
                byte[] data = File.ReadAllBytes(filePath);
                using (MemoryStream ms = new MemoryStream(data))
                using (BinaryReader reader = new BinaryReader(ms))
                {
                    // Skip magic number
                    reader.ReadByte();

                    // Read header
                    DemoHeader header = new DemoHeader();
                    reader.ReadByte(); // magic number
                    header.formatVersion = reader.ReadInt32();
                    header.gameVersion = ReadString(reader);
                    header.timestamp = reader.ReadInt64();
                    header.description = ReadString(reader);
                    header.totalFrames = reader.ReadUInt32();
                    header.projectileEvents = reader.ReadUInt32();
                    header.targetEvents = reader.ReadUInt32();
                    header.duration = reader.ReadSingle();
                    header.checksum = reader.ReadUInt32();
                    header.playerStartPosition = ReadVector3(reader);
                    header.playerStartRotation = ReadQuaternion(reader);
                    header.playerStartVelocity = ReadVector3(reader);

                    return new DemoInfo
                    {
                        filePath = filePath,
                        fileName = fileInfo.Name,
                        fileSize = fileInfo.Length,
                        timestamp = header.timestamp,
                        mapName = "Unknown",
                        gameMode = "Unknown",
                        duration = header.duration,
                        frameCount = (int)header.totalFrames,
                        formatVersion = (byte)header.formatVersion,
                        gameVersion = header.gameVersion,
                        recordedInputs = false
                    };
                }
            }
            catch
            {
                return null;
            }
        }

        /// <summary>
        /// Save a demo file.
        /// </summary>
        public string SaveDemo(DemoFile demo, string customName = null)
        {
            if (demo == null)
            {
                Debug.LogError("[DemoManager] Cannot save null demo");
                return null;
            }

            // Generate filename
            string fileName = customName ?? GenerateDemoFileName(demo);
            string filePath = Path.Combine(_fullDemosPath, fileName);

            // Save the file
            try
            {
                DemoSerializer.SaveToFile(demo, filePath);

                // Add to cache
                DemoInfo info = LoadDemoInfo(filePath);
                if (info != null)
                {
                    _demoCache.Add(info);
                    _demoCache = _demoCache.OrderByDescending(d => d.timestamp).ToList();

                    OnDemoSaved?.Invoke(info);
                    OnDemoCacheUpdated?.Invoke();
                }

                // Check storage limits
                CheckStorageLimits();

                Debug.Log($"[DemoManager] Saved demo: {fileName}");
                return filePath;
            }
            catch (Exception e)
            {
                Debug.LogError($"[DemoManager] Failed to save demo: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// Save current recording from DemoRecorder.
        /// </summary>
        public string SaveCurrentRecording(string customName = null)
        {
            if (_recorder == null)
            {
                Debug.LogError("[DemoManager] No DemoRecorder found");
                return null;
            }

            DemoFile demo = _recorder.ExtractDemoFile();
            return SaveDemo(demo, customName);
        }

        /// <summary>
        /// Load a demo file.
        /// </summary>
        public DemoFile LoadDemo(string filePath)
        {
            if (!File.Exists(filePath))
            {
                Debug.LogError($"[DemoManager] Demo file not found: {filePath}");
                return null;
            }

            try
            {
                DemoFile demo = DemoSerializer.LoadFromFile(filePath);
                Debug.Log($"[DemoManager] Loaded demo: {Path.GetFileName(filePath)}");
                return demo;
            }
            catch (Exception e)
            {
                Debug.LogError($"[DemoManager] Failed to load demo: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// Load a demo by index and load it into the DemoPlayer.
        /// </summary>
        public bool LoadAndPlayDemo(int index)
        {
            if (index < 0 || index >= _demoCache.Count)
            {
                Debug.LogError($"[DemoManager] Invalid demo index: {index}");
                return false;
            }

            DemoInfo info = _demoCache[index];
            DemoFile demo = LoadDemo(info.filePath);

            if (demo == null)
                return false;

            if (_player != null)
            {
                _player.LoadDemo(demo);
                _player.Play();
                return true;
            }

            Debug.LogError("[DemoManager] No DemoPlayer found");
            return false;
        }

        /// <summary>
        /// Delete a demo file.
        /// </summary>
        public bool DeleteDemo(string filePath)
        {
            if (!File.Exists(filePath))
            {
                Debug.LogWarning($"[DemoManager] Demo file not found: {filePath}");
                return false;
            }

            try
            {
                File.Delete(filePath);

                // Remove from cache
                DemoInfo info = _demoCache.FirstOrDefault(d => d.filePath == filePath);
                if (info != null)
                {
                    _demoCache.Remove(info);
                    OnDemoDeleted?.Invoke(info);
                    OnDemoCacheUpdated?.Invoke();
                }

                Debug.Log($"[DemoManager] Deleted demo: {Path.GetFileName(filePath)}");
                return true;
            }
            catch (Exception e)
            {
                Debug.LogError($"[DemoManager] Failed to delete demo: {e.Message}");
                return false;
            }
        }

        /// <summary>
        /// Delete a demo by index.
        /// </summary>
        public bool DeleteDemo(int index)
        {
            if (index < 0 || index >= _demoCache.Count)
                return false;

            return DeleteDemo(_demoCache[index].filePath);
        }

        /// <summary>
        /// Get demo info by index.
        /// </summary>
        public DemoInfo GetDemoInfo(int index)
        {
            if (index < 0 || index >= _demoCache.Count)
                return null;

            return _demoCache[index];
        }

        /// <summary>
        /// Get all demo infos.
        /// </summary>
        public DemoInfo[] GetAllDemoInfos()
        {
            return _demoCache.ToArray();
        }

        /// <summary>
        /// Check and enforce storage limits.
        /// </summary>
        private void CheckStorageLimits()
        {
            if (!autoDeleteOldDemos)
                return;

            // Delete old demos by date
            if (demoRetentionDays > 0)
            {
                long cutoffTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds() - (demoRetentionDays * 24 * 60 * 60);

                var oldDemos = _demoCache.Where(d => d.timestamp < cutoffTime).ToList();
                foreach (var demo in oldDemos)
                {
                    DeleteDemo(demo.filePath);
                }
            }

            // Delete oldest demos if over count limit
            if (maxDemosToKeep > 0 && _demoCache.Count > maxDemosToKeep)
            {
                int toDelete = _demoCache.Count - maxDemosToKeep;
                for (int i = 0; i < toDelete; i++)
                {
                    DeleteDemo(_demoCache.Count - 1);  // Delete oldest (at end of sorted list)
                }
            }

            // Delete oldest demos if over storage limit
            if (maxStorageBytes > 0 && TotalStorageUsed > maxStorageBytes)
            {
                while (TotalStorageUsed > maxStorageBytes && _demoCache.Count > 0)
                {
                    DeleteDemo(_demoCache.Count - 1);  // Delete oldest
                }
            }
        }

        /// <summary>
        /// Calculate total storage used by demos.
        /// </summary>
        private long CalculateTotalStorage()
        {
            return _demoCache.Sum(d => d.fileSize);
        }

        /// <summary>
        /// Generate a filename for a demo file.
        /// </summary>
        private string GenerateDemoFileName(DemoFile demo)
        {
            DateTime dateTime = DateTimeOffset.FromUnixTimeSeconds(demo.header.timestamp).DateTime;
            string timestamp = dateTime.ToString("yyyyMMdd_HHmmss");
            return $"demo_{timestamp}.demo";
        }

        /// <summary>
        /// Helper to read string from binary reader.
        /// </summary>
        private string ReadString(BinaryReader reader)
        {
            ushort length = reader.ReadUInt16();
            if (length == 0)
                return string.Empty;

            byte[] bytes = reader.ReadBytes(length);
            return System.Text.Encoding.UTF8.GetString(bytes);
        }

        private static Vector3 ReadVector3(BinaryReader reader)
        {
            return new Vector3(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());
        }

        private static Quaternion ReadQuaternion(BinaryReader reader)
        {
            return new Quaternion(reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle());
        }

        /// <summary>
        /// Demo metadata information.
        /// </summary>
        [Serializable]
        public class DemoInfo
        {
            public string filePath;
            public string fileName;
            public long fileSize;
            public long timestamp;
            public string mapName;
            public string gameMode;
            public float duration;
            public int frameCount;
            public byte formatVersion;
            public string gameVersion;
            public bool recordedInputs;

            public string FileSizeMB => (fileSize / (1024f * 1024f)).ToString("F2");
            public string FormattedDate => DateTimeOffset.FromUnixTimeSeconds(timestamp).ToString("yyyy-MM-dd HH:mm");
        }
    }
}

File: client/src/zzz_demo_zzz/management/DemoManager.ts
import { DemoFile } from '../types/DemoFile.js';
import { DemoSerializer } from '../serialization/DemoSerializer.js';
import { VersionManager } from '../serialization/VersionManager.js';
import { SchemaRegistry } from '../serialization/SchemaRegistry.js';

/**
 * Demo file metadata.
 */
export interface DemoInfo {
  /** File path or ID */
  id: string;
  /** Demo name */
  name: string;
  /** File size in bytes */
  size: number;
  /** Recording timestamp */
  timestamp: number;
  /** Duration in seconds */
  duration: number;
  /** Number of frames */
  frameCount: number;
  /** Number of projectile events */
  projectileEvents: number;
  /** Number of target events */
  targetEvents: number;
  /** Description */
  description: string;
  /** Format version */
  formatVersion: number;
}

/**
 * Manages demo file operations.
 * Handles saving, loading, browsing, and cleanup of demo files.
 * Uses IndexedDB for browser storage.
 */
export class DemoManager {
  private schemaRegistry: SchemaRegistry;
  private versionManager: VersionManager;
  private demoCache: DemoInfo[] = [];
  private dbName: string = 'DemoStorage';
  private storeName: string = 'demos';
  private db: IDBDatabase | null = null;

  // Configuration
  private maxDemosToKeep: number = 50;
  private maxStorageBytes: number = 50 * 1024 * 1024; // 50 MB
  private autoDeleteOldDemos: boolean = true;
  private demoRetentionDays: number = 30;

  constructor(schemaRegistry?: SchemaRegistry) {
    this.schemaRegistry = schemaRegistry || new SchemaRegistry();
    this.versionManager = new VersionManager(this.schemaRegistry);
  }

  /**
   * Initialize the demo manager and open IndexedDB.
   */
  async initialize(): Promise<void> {
    return new Promise((resolve, reject) => {
      const request = indexedDB.open(this.dbName, 1);

      request.onerror = () => {
        reject(new Error(`Failed to open IndexedDB: ${request.error}`));
      };

      request.onsuccess = () => {
        this.db = request.result;
        console.log('[DemoManager] IndexedDB initialized');
        this.refreshDemoCache().then(resolve).catch(reject);
      };

      request.onupgradeneeded = (event) => {
        const db = (event.target as IDBOpenDBRequest).result;
        if (!db.objectStoreNames.contains(this.storeName)) {
          const store = db.createObjectStore(this.storeName, { keyPath: 'id' });
          store.createIndex('timestamp', 'timestamp', { unique: false });
          store.createIndex('name', 'name', { unique: false });
        }
      };
    });
  }

  /**
   * Save a demo file to storage.
   * @param demo - Demo file to save
   * @param customName - Optional custom name
   * @returns Demo info
   */
  async saveDemo(demo: DemoFile, customName?: string): Promise<DemoInfo> {
    if (!this.db) {
      throw new Error('DemoManager not initialized');
    }

    // Serialize demo
    const data = DemoSerializer.serialize(demo);
    const name = customName || `demo_${Date.now()}`;

    // Create demo info
    const demoInfo: DemoInfo = {
      id: name,
      name,
      size: data.length,
      timestamp: demo.header.timestamp,
      duration: demo.header.duration,
      frameCount: demo.header.totalFrames,
      projectileEvents: demo.header.projectileEvents,
      targetEvents: demo.header.targetEvents,
      description: demo.header.description,
      formatVersion: demo.header.formatVersion,
    };

    // Save to IndexedDB
    return new Promise((resolve, reject) => {
      const transaction = this.db!.transaction([this.storeName], 'readwrite');
      const store = transaction.objectStore(this.storeName);

      const record = {
        id: demoInfo.id,
        name: demoInfo.name,
        data,
        info: demoInfo,
      };

      const request = store.put(record);

      request.onsuccess = () => {
        console.log(`[DemoManager] Saved demo: ${name}`);
        this.refreshDemoCache();
        this.checkStorageLimits();
        resolve(demoInfo);
      };

      request.onerror = () => {
        reject(new Error(`Failed to save demo: ${request.error}`));
      };
    });
  }

  /**
   * Load a demo file from storage.
   * @param id - Demo ID
   * @returns Demo file
   */
  async loadDemo(id: string): Promise<DemoFile> {
    if (!this.db) {
      throw new Error('DemoManager not initialized');
    }

    return new Promise((resolve, reject) => {
      const transaction = this.db!.transaction([this.storeName], 'readonly');
      const store = transaction.objectStore(this.storeName);
      const request = store.get(id);

      request.onsuccess = () => {
        const record = request.result;
        if (!record) {
          reject(new Error(`Demo not found: ${id}`));
          return;
        }

        // Deserialize demo
        let demo = DemoSerializer.deserialize(record.data);

        // Migrate to current version if needed
        if (this.versionManager.needsMigration(demo)) {
          console.log(`[DemoManager] Migrating demo from v${demo.header.formatVersion}`);
          demo = this.versionManager.migrateToCurrent(demo);
        }

        resolve(demo);
      };

      request.onerror = () => {
        reject(new Error(`Failed to load demo: ${request.error}`));
      };
    });
  }

  /**
   * Delete a demo file from storage.
   * @param id - Demo ID
   */
  async deleteDemo(id: string): Promise<void> {
    if (!this.db) {
      throw new Error('DemoManager not initialized');
    }

    return new Promise((resolve, reject) => {
      const transaction = this.db!.transaction([this.storeName], 'readwrite');
      const store = transaction.objectStore(this.storeName);
      const request = store.delete(id);

      request.onsuccess = () => {
        console.log(`[DemoManager] Deleted demo: ${id}`);
        this.refreshDemoCache();
        resolve();
      };

      request.onerror = () => {
        reject(new Error(`Failed to delete demo: ${request.error}`));
      };
    });
  }

  /**
   * Get all demo info.
   * @returns Array of demo info
   */
  getDemos(): DemoInfo[] {
    return [...this.demoCache];
  }

  /**
   * Get demo info by ID.
   * @param id - Demo ID
   * @returns Demo info or undefined
   */
  getDemoInfo(id: string): DemoInfo | undefined {
    return this.demoCache.find((demo) => demo.id === id);
  }

  /**
   * Refresh the demo cache from storage.
   */
  private async refreshDemoCache(): Promise<void> {
    if (!this.db) {
      return;
    }

    return new Promise((resolve, reject) => {
      const transaction = this.db!.transaction([this.storeName], 'readonly');
      const store = transaction.objectStore(this.storeName);
      const request = store.getAll();

      request.onsuccess = () => {
        const records = request.result;
        this.demoCache = records
          .map((record: any) => record.info)
          .sort((a: DemoInfo, b: DemoInfo) => b.timestamp - a.timestamp);
        resolve();
      };

      request.onerror = () => {
        reject(new Error(`Failed to refresh demo cache: ${request.error}`));
      };
    });
  }

  /**
   * Check and enforce storage limits.
   */
  private async checkStorageLimits(): Promise<void> {
    if (!this.autoDeleteOldDemos) {
      return;
    }

    // Check count limit
    if (this.demoCache.length > this.maxDemosToKeep) {
      const toDelete = this.demoCache.slice(this.maxDemosToKeep);
      for (const demo of toDelete) {
        await this.deleteDemo(demo.id);
      }
    }

    // Check size limit
    const totalSize = this.demoCache.reduce((sum, demo) => sum + demo.size, 0);
    if (totalSize > this.maxStorageBytes) {
      // Delete oldest demos until under limit
      let sorted = [...this.demoCache].sort((a, b) => a.timestamp - b.timestamp);
      let currentSize = totalSize;

      while (currentSize > this.maxStorageBytes && sorted.length > 0) {
        const oldest = sorted.shift()!;
        await this.deleteDemo(oldest.id);
        currentSize -= oldest.size;
      }
    }

    // Check retention limit
    const cutoffTime = Date.now() - this.demoRetentionDays * 24 * 60 * 60 * 1000;
    const expired = this.demoCache.filter((demo) => demo.timestamp < cutoffTime);
    for (const demo of expired) {
      await this.deleteDemo(demo.id);
    }
  }

  /**
   * Get total storage used.
   * @returns Total size in bytes
   */
  getTotalStorageUsed(): number {
    return this.demoCache.reduce((sum, demo) => sum + demo.size, 0);
  }

  /**
   * Get storage usage percentage.
   * @returns Percentage (0-100)
   */
  getStorageUsagePercentage(): number {
    return (this.getTotalStorageUsed() / this.maxStorageBytes) * 100;
  }

  /**
   * Clear all demos.
   */
  async clearAllDemos(): Promise<void> {
    if (!this.db) {
      throw new Error('DemoManager not initialized');
    }

    return new Promise((resolve, reject) => {
      const transaction = this.db!.transaction([this.storeName], 'readwrite');
      const store = transaction.objectStore(this.storeName);
      const request = store.clear();

      request.onsuccess = () => {
        console.log('[DemoManager] Cleared all demos');
        this.refreshDemoCache();
        resolve();
      };

      request.onerror = () => {
        reject(new Error(`Failed to clear demos: ${request.error}`));
      };
    });
  }

  // Configuration setters
  setMaxDemosToKeep(max: number): void {
    this.maxDemosToKeep = max;
  }

  setMaxStorageBytes(max: number): void {
    this.maxStorageBytes = max;
  }

  setAutoDeleteOldDemos(auto: boolean): void {
    this.autoDeleteOldDemos = auto;
  }

  setDemoRetentionDays(days: number): void {
    this.demoRetentionDays = days;
  }

  /**
   * Close the IndexedDB connection.
   */
  close(): void {
    if (this.db) {
      this.db.close();
      this.db = null;
    }
  }
}
